internal/cache/base.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// ForceCleanFor forces a complete remove of all the objects that belong to the specified
+// canvas. Usually used to free all objects from a closing windows.
+func ForceCleanFor(canvas fyne.Canvas) {
+	deletingObjs := make([]fyne.CanvasObject, 0, 50)
+
+	// find all objects that belong to the specified canvas
+	canvasesLock.RLock()
+	for obj, cinfo := range canvases {
+		if cinfo.canvas == canvas {
+			deletingObjs = append(deletingObjs, obj)
+		}
+	}
+	canvasesLock.RUnlock()
+	if len(deletingObjs) == 0 {
+		return
+	}
+
+	// remove them from canvas cache
+	canvasesLock.Lock()
+	for _, dobj := range deletingObjs {
+		delete(canvases, dobj)
+	}
+	canvasesLock.Unlock()
+
+	// destroy their renderers and delete them from the renderer
+	// cache if they are widgets
+	renderersLock.Lock()
+	for _, dobj := range deletingObjs {
+		wid, ok := dobj.(fyne.Widget)
+		if !ok {
+			continue
+		}
+		winfo, ok := renderers[wid]
+		if !ok {
+			continue
+		}
+		winfo.renderer.Destroy()
+		delete(renderers, wid)
+	}
+	renderersLock.Unlock()
+}
+
+// NotifyCanvasRefresh notifies to the caches that a canvas refresh was triggered.
+func NotifyCanvasRefresh() {
+	select {
+	case canvasRefreshCh <- struct{}{}:
+	default:
+		return
+	}
+}
+
+// ===============================================================
+// Private functions
+// ===============================================================
+
+// destroyExpiredCanvases deletes objects from the canvases cache.
+func destroyExpiredCanvases(now time.Time) {
+	expiredObjects = expiredObjects[:0]
+	canvasesLock.RLock()
+	for obj, cinfo := range canvases {
+		if cinfo.isExpired(now) {
+			expiredObjects = append(expiredObjects, obj)
+		}
+	}
+	canvasesLock.RUnlock()
+	if len(expiredObjects) > 0 {
+		canvasesLock.Lock()
+		for i, exp := range expiredObjects {
+			delete(canvases, exp)
+			expiredObjects[i] = nil
+		}
+		canvasesLock.Unlock()
+	}
+}
+
+// destroyExpiredRenderers deletes the renderer from the cache and calls
+// renderer.Destroy()
+func destroyExpiredRenderers(now time.Time) {
+	expiredObjects = expiredObjects[:0]
+	renderersLock.RLock()
+	for wid, rinfo := range renderers {
+		if rinfo.isExpired(now) {
+			rinfo.renderer.Destroy()
+			expiredObjects = append(expiredObjects, wid)
+		}
+	}
+	renderersLock.RUnlock()
+	if len(expiredObjects) > 0 {
+		renderersLock.Lock()
+		for i, exp := range expiredObjects {
+			delete(renderers, exp.(fyne.Widget))
+			expiredObjects[i] = nil
+		}
+		renderersLock.Unlock()
+	}
+}
+
+// destroyExpiredSvgs destroys expired svgs cache data.
+func destroyExpiredSvgs(now time.Time) {
+	expiredSvgs := make([]string, 0, 20)
+	svgLock.RLock()
+	for s, sinfo := range svgs {
+		if sinfo.isExpired(now) {
+			expiredSvgs = append(expiredSvgs, s)
+		}
+	}
+	svgLock.RUnlock()
+	if len(expiredSvgs) > 0 {
+		svgLock.Lock()
+		for _, exp := range expiredSvgs {
+			delete(svgs, exp)
+		}
+		svgLock.Unlock()
+	}
+}
+
+// ===============================================================
```

Comment:
As above, these comments don't add meaning, if it's required for readability maybe we need a new file.

Reply:
done in ad7f974